### PR TITLE
Polished up pretty much everything

### PR DESCRIPTION
**Pull request by sylvinite**
Refactoring:
- Reworked templates for providing AWS credentials
- Refactored the code. Squished functions together, reorganised files
- Seperated HCI and non HCI dependent commands
- Restructed the CLI stack
- Unobfuscated the pretty variable

Improvements:
- Rewrote argparse into click support
- Rewrote structure into installable package
- Replaced print with global logging functions
- Wrote self-test for 61% code coverage
- CLI can now call up AWS by using a credentials file
- Added automatic connection test when creating HPC object
- Possible to upload folder structures containing fastq
- Crap queries (search,delete) will try to resolve situation anyway

Fresh stuff:
- Added license
- Added conda autoinstaller
- Wrote an attempt for TravisCI autobuild
- Added a couple of fastq file verification steps
- Added function to auto-generate pipeline tags of uploaded fastq
- Added versioning

### CODE REVIEW

File: HCPInterface/cli/base.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+@click.option("-id","--access_key_id",help="Amazon key identifier",type=str,default="")
+@click.option("-key","--access_key",help="Amazon secret access key",type=str,default="")
+@click.option('-c',"--credentials", help="File containing ep, id & key; instead of using the CLI.",type=str,default="")
+@click.option("-b","--bucket",help="Bucket name",type=str, required=True)
+@click.version_option(version)
+@click.pass_context
+def root(ctx, endpoint, access_key_id, access_key, bucket, credentials):
+    """HCP interfacing tool"""
+    #Invalid input checks
+    if credentials != "" and (endpoint != "" or access_key_id != "" or access_key != ""):
```
Comment (JPWerling) [inline HCPInterface/cli/base.py]: `if credentials and any([endpoint, access_key_id, access_key]):`
Readability

File: HCPInterface/cli/base.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        log.error("Credentials were provided both through a file and the CLI. Make up your mind")
+        sys.exit(-1)
```
Comment (JPWerling) [inline HCPInterface/cli/base.py]: Why `sys` over `click.abort()`?

File: HCPInterface/hcp/hcp.py [hunk 3]
```diff
@@ -17,10 +18,12 @@
...
 from boto3.s3.transfer import TransferConfig
 
-from .helpers import calculate_etag
-from .errors import (UnattachedBucketError, LocalFileExistsError,
-                     UnknownSourceTypeError, MismatchChecksumError)
-from .config import get_config
+from HCPInterface.hcp.helpers import calculate_etag
+from HCPInterface.hcp.errors import (UnattachedBucketError, LocalFileExistsError,
+                     UnknownSourceTypeError, MismatchChecksumError, ConnectionError,
+                     MissingCredentialsError)
```
Comment (JPWerling) [inline HCPInterface/hcp/hcp.py]: Indentation

File: HCPInterface/hcp/hcp.py [hunk 4]
```diff
@@ -114,16 +121,41 @@ def __init__(self, endpoint, aws_access_key_id, aws_secret_access_key, debug=Fal
...
+
     def list_buckets(self):
         """List all available buckets at endpoint."""
         return [bucket.name for bucket in self.s3.buckets.all()]
 
+    def test_connection(self):
+        """Validate the connection works with as little overhead as possible."""
+        try:
+            if self.bucket is None:
+                raise ConnectionError("Unassigned bucket")
```
Comment (JPWerling) [inline HCPInterface/hcp/hcp.py]: Unattached bucket rather than unassigned

File: HCPInterface/hcp/hcp.py [hunk 5]
```diff
@@ -158,17 +190,23 @@ def search_objects(self, string):
         return [obj for obj in self.objects if string in obj.key]
 
     @bucketcheck
-    def upload_file(self, local_path, remote_key, metadata={}):
+    def upload_file(self, local_path, remote_key, metadata={}, force=False):
```
Comment (JPWerling) [inline HCPInterface/hcp/hcp.py]: Force has been intentionally left out from upload functionality due to risk of overwriting clinical data. The argument made was should the need arise to remove erroneous data then it should be manually (and therefore fully intentionally) deleted prior to uploading via the normal means.

Whether this still stands is a topic of discussion.

File: HCPInterface/hcp/hcp.py [hunk 6]
```diff
@@ -158,17 +190,23 @@ def search_objects(self, string):
...
         self.bucket.upload_file(local_path,
                                 remote_key,
                                 ExtraArgs={'Metadata': metadata},
                                 Config=self.transfer_config,
                                 Callback=ProgressPercentage(local_path))
         print('')  # Post progressbar correction for stdout
 
-        calculated_etag = calculate_etag(local_path)
         remote_obj = self.get_object(remote_key)
+        calculated_etag = calculate_etag(local_path)
```
Comment (JPWerling) [inline HCPInterface/hcp/hcp.py:228]: Reason for moving this?

File: freeze.txt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+sh==1.14.1
+six==1.16.0
+toml==0.10.2
+tomli==1.2.1
+typed-ast==1.4.3
+typing-extensions==3.10.0.2
+urllib3==1.25.11
+wrapt==1.12.1
+yamllint==1.26.3
+zipp==3.5.0
```
Comment (JPWerling) [inline freeze.txt:46]: Why push output from `pip freeze`?

**Review by JD2112 — APPROVED** [review]
For me this is pretty cool code and ready to merge

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+except IOError:
+    install_requires = []
+
+setup(
+    name="HCPInterface",
+    version=version,
+    long_description=__doc__,
+    url="https://github.com/genomic-medicine-sweden/HCPInterface",
+    author="Isak Sylvin",
+    author_email="isak.sylvin@gu.se",
```
Comment (JPWerling) [inline setup.py:17]: Set clinicalgenomics@gu.se?

File: setup.sh [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+scriptdir=`dirname $SCRIPT`
+NAME=${1:-hcpenv}
+
+echo "INFO - Unloading active environment"
+conda info| tac | tac | grep -q $NAME && source deactivate || :
+echo "INFO - Removing existing instance of environment (if present)"
+conda remove -y -n $NAME --all || :
+
+echo "INFO - Creating new environment"
+conda create --yes --name $NAME python=3.7
```
Comment (JPWerling) [inline setup.sh:13]: Previous reference to python version in travis lists python=3.6

File: setup.sh [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+source activate $NAME
+
+echo "INFO - Installing requirements"
+conda config --add channels bioconda
+pip install -r ${scriptdir}/requirements.txt
+pip install -r ${scriptdir}/freeze.txt
```
Comment (JPWerling) [inline setup.sh:19]: freeze.txt seems unnecessary to include

File: tests/test_hcp.py [hunk 11]
```diff
@@ -9,36 +9,20 @@
...
-if not all([endpoint, aws_access_key_id, aws_secret_access_key]):
-    raise MissingCredentialsError('One or more credentials missing from keys.json.')
-
-
-hcpm = HCPManager(endpoint, aws_access_key_id, aws_secret_access_key)
+hcpm = HCPManager(credentials_path=credentials_path)
 hcpm.attach_bucket("ngs-test")
 
-
 class TestProcess(unittest.TestCase):
```
Comment (JPWerling) [inline tests/test_hcp.py:22]: These tests pose a problem in general. Each test should be isolated and not depend on the previous one. Should a crash occur before deletion on the remote then subsequent tests will malfunction. The tests do not work for other sites either as we point directly at our bucket ngs-test.

Comment (JPWerling) [reply tests/test_hcp.py:22]: In addition, when uploading to a key with a folder structure, an object is made on each level. In other words, temp/file.txt generates 2 object, one with key temp/ and one with temp/file.txt. 

The tests only removes the latter object.

File: tests/test_cmds.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    #Command should complain about lack of subcommand
+    assert res.exit_code == 2
+
+def test_hci_base(runner):
+    cmd = "-b ngs-test -c {0} hci".format(credentials_path)
+    res = runner.invoke(root, cmd.split())
+    assert res.exit_code == 0
+
+def test_upload(runner):
+    cmd = "-b ngs-test -c {0} upload -i {1} -d {2} -f -m /tmp/meta.json".format(credentials_path,
```
Comment (JPWerling) [inline tests/test_cmds.py]: is `/tmp/meta.json` generated dynamically on test runs?

File: HCPInterface/io/io.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    if not hit:
+        raise Exception("File {0} is not a zipped fastq".format(fn))
+    log.debug('Verified that {0} is a zipped fastq'.format(fn))
+
+def verify_fq_content(fn):
+    """Makes sure fastq file contains fastq data"""
+    nuc = set("ATCG\n")
+    lineno = 0
+    f1 = gzip.open(fn, "r")
+    for line in f1:
```
Comment (JPWerling) [inline HCPInterface/io/io.py:32]: Does this parse the entire fastq?

File: HCPInterface/io/io.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    hit = re.search("fastq.gz$",fn)
+    if not hit:
+        raise Exception("File {0} is not a zipped fastq".format(fn))
+    log.debug('Verified that {0} is a zipped fastq'.format(fn))
+
+def verify_fq_content(fn):
+    """Makes sure fastq file contains fastq data"""
+    nuc = set("ATCG\n")
+    lineno = 0
+    f1 = gzip.open(fn, "r")
```
Comment (JPWerling) [inline HCPInterface/io/io.py:31]: This open never closes

File: HCPInterface/io/io.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+def verify_fq_suffix(fn):
+    """Makes sure the provided file has fastq suffic"""
+    #Check suffix
+    hit = re.search("fastq.gz$",fn)
+    if not hit:
+        raise Exception("File {0} is not a zipped fastq".format(fn))
+    log.debug('Verified that {0} is a zipped fastq'.format(fn))
+
+def verify_fq_content(fn):
```
Comment (JPWerling) [inline HCPInterface/io/io.py:27]: Consider using biopython
```
from Bio import SeqIO
for record in SeqIO.parse(<input>, "fastq"):
    pass
```

File: HCPInterface/io/io.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+import json
+import os
+import re
+
+from HCPInterface import log, WD, TIMESTAMP
+
+def verify_fq_suffix(fn):
+    """Makes sure the provided file has fastq suffic"""
+    #Check suffix
+    hit = re.search("fastq.gz$",fn)
```
Comment (JPWerling) [inline HCPInterface/io/io.py]: Do all fastq have to be compressed?

Comment (JPWerling) [reply HCPInterface/io/io.py]: Output from Torrent Machines for example are not gzipped by default.

File: HCPInterface/io/io.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+            corr = re.match("^@",line)
+        elif lineno == 2:
+            corr = set(line) <= nuc
+        elif lineno == 3:
+            corr = re.match("^\+",line)
+        elif lineno == 4:
+            lineno = 0
+
+        if not corr:
+            raise Exception("File {0} does not look like a fastq at line {1}: {2}".format(fn, lineno, line))
```
Comment (JPWerling) [inline HCPInterface/io/io.py]: f-strings are beautiful

File: HCPInterface/io/io.py [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+    if lineno % 4 != 0:
+        raise Exception("File {0} is missing data".format(fn))
+    log.debug('Verified that {0} resembles a fastq'.format(fn))
+
+def generate_tagmap(fn, tag, out="{}/meta-{}.json".format(os.getcwd(), TIMESTAMP)):
+    """Creates a json file with filenames and tags"""
+    mdict = dict()
+    mdict[fn] = {'tag':tag}
+    md = open(out, "a")
```
Comment (JPWerling) [inline HCPInterface/io/io.py:56]: Context manager `with open() as out:` for automatic closing

File: HCPInterface/hcp/hcp.py (hunk 4 above)
Comment (sylvinite) [reply HCPInterface/hcp/hcp.py]: Nah. As a bucket hasn't been provided, no bucket has been assigned to the variable to attach it

File: HCPInterface/hcp/hcp.py (hunk 5 above)
Comment (sylvinite) [reply HCPInterface/hcp/hcp.py]: Imported comment for for future clarification

File: HCPInterface/io/io.py (hunk 15 above)
Comment (sylvinite) [reply HCPInterface/io/io.py:27]: Feh. Adds complexity and there is no gain currently. Taking a mental note if we wanna make it even shinier in the future

File: HCPInterface/io/io.py (hunk 17 above)
Comment (sylvinite) [reply HCPInterface/io/io.py]: Replaced all format invocations with fstrings

File: HCPInterface/io/io.py (hunk 18 above)
Comment (sylvinite) [reply HCPInterface/io/io.py:56]: I like the explicit "we are only using this for a single command" thou

**Review by JPWerling — APPROVED** [review]

**Comment by JPWerling** [comment]
Consider rebase on dev prior to merge. Likely not needed however.

**Pull request merged**